### PR TITLE
Closes #2295 - Implement EDITTASKS Permission

### LINKED ISSUE #2295: Implement EDITTASKS Permission
This issue is related with #2242

### PR DESCRIPTION
**Pull request by jamesrdi**
https://sonarcloud.io/summary/new_code?id=jamesrdi_taskana&branch=TSK-2295

---
Release Notes:

```

```
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [ ] After integration of the PR, I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [ ] Commit message format → (Closes) #<Issue Number>: Your commit message. i.e. Closes #2271: Your commit message.
- [ ] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [ ] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

**Review by ryzheboka — CHANGES_REQUESTED** [review]
Great work :tada: Could you add a link to Sonarcloud?

**Comment by ryzheboka** [comment]
Please rebase so that you have only one commit :)

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java [hunk 1]
```diff
@@ -624,11 +625,24 @@ public void forceDeleteTask(String taskId)
   }
 
   @Override
-  public Optional<Task> selectAndClaim(TaskQuery taskQuery) {
+  public Optional<Task> selectAndClaim(TaskQuery taskQuery)
+      throws NotAuthorizedOnWorkbasketException {
+    // Temporary Fix because of Oracle, the queryTaskSummariesOracle() does not prove the
+    // authorization, when single() is called without READ permission, query still returns a task
+    // although it should not, and TASKANA tries to append a Workbasket to this Task and fails
+    // because of missing permission
```
Comment (ryzheboka) [inline lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java:633]: We don't need this comment anymore after the change in logic

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java [hunk 2]
```diff
@@ -704,9 +724,25 @@ public List<String> updateTasks(
       // use query in order to find only those tasks that are visible to the current user
       List<TaskSummary> taskSummaries = getTasksToChange(selectionCriteria);
 
+      List<TaskSummary> taskWithPermissions = new ArrayList<>();
+      for (TaskSummary taskSummary : taskSummaries) {
+        WorkbasketQueryImpl query = (WorkbasketQueryImpl) workbasketService.createWorkbasketQuery();
+        String workbasketId = taskSummary.getWorkbasketSummary().getId();
+        WorkbasketSummary workbasket =
```
[the remark this answers is not in the document]
Comment (ryzheboka) [reply lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java:725]: Why don't you use checkEditTasksPerm here?

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java [hunk 3]
```diff
@@ -738,9 +775,25 @@ public List<String> updateTasks(
...
+            query
+                .idIn(workbasketId)
+                .callerHasPermissions(
+                    WorkbasketPermission.READTASKS, WorkbasketPermission.EDITTASKS)
```
[the remark this answers is not in the document]
Comment (ryzheboka) [reply lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java]: Why don't you use checkEditTasksPerm here?

**Review by ryzheboka — CHANGES_REQUESTED** [review]

File: lib/taskana-core-test/src/test/java/acceptance/workbasket/get/GetWorkbasketAccTest.java [hunk 4]
```diff
@@ -103,11 +103,12 @@ void should_ReturnWorkbasketPermissions_When_IdIsValidAndUserHasPermissions() {
     List<WorkbasketPermission> permissions =
         workbasketService.getPermissionsForWorkbasket(defaultWorkbasketSummary.getId());
 
-    assertThat(permissions).hasSize(4);
-    assertThat(permissions.contains(WorkbasketPermission.READ)).isTrue();
-    assertThat(permissions.contains(WorkbasketPermission.OPEN)).isTrue();
-    assertThat(permissions.contains(WorkbasketPermission.TRANSFER)).isTrue();
-    assertThat(permissions.contains(WorkbasketPermission.APPEND)).isTrue();
+    assertThat(permissions)
```
Comment (ryzheboka) [inline lib/taskana-core-test/src/test/java/acceptance/workbasket/get/GetWorkbasketAccTest.java:106]: Please change assertion to "containsExactlyInAnyOrder". You don't need to check the size.

**Review by holgerhagen — CHANGES_REQUESTED** [review]

**Review by gitgoodjhe — APPROVED** [review]

**Review by ryzheboka — APPROVED** [review]

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java [hunk 5]
```diff
@@ -1552,6 +1599,12 @@ private Task completeTask(String taskId, boolean isForced)
         return task;
       }
 
+      if (!checkEditTasksPerm(task)) {
```
Comment (holgerhagen) [inline lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java:1602]: Isn't this already checked in "checkPreconditions"?

**Review by holgerhagen — CHANGES_REQUESTED** [review]

**Review by holgerhagen — APPROVED** [review]

**Pull request merged**